near-sdk/src/store/free_list/mod.rs

```diff
@@ -220,6 +220,86 @@ where
     pub fn drain(&mut self) -> Drain<T> {
         Drain::new(self)
     }
+
+    /// Empty slots in the front of the list is swapped with occupied slots in back of the list.
+    /// Defrag helps reduce gas cost in certain scenarios where lot of elements in front of the list are
+    /// removed without getting replaced. Please see https://github.com/near/near-sdk-rs/issues/990
+    pub fn defrag<F>(&mut self, callback: F)
+    where
+        F: FnMut(&T, u32),
+    {
+        Defrag::new(self).defrag(callback);
+        self.first_free = None;
+    }
+}
+
+///Defrag struct has helper functions to perform defragmentation of `FreeList`. See the
+/// documentation of function `FreeList->defrag` for more details.
+struct Defrag<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    elements: &'a mut Vector<Slot<T>>,
+    occupied_count: u32,
+    curr_free_slot: Option<FreeListIndex>,
+    defrag_index: u32,
+}
+
+impl<'a, T> Defrag<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    /// Create a new struct for defragmenting `FreeList`.
+    fn new(list: &'a mut FreeList<T>) -> Self {
+        Self {
+            elements: &mut list.elements,
+            occupied_count: list.occupied_count,
+            defrag_index: list.occupied_count,
+            curr_free_slot: list.first_free,
+        }
+    }
+
+    fn defrag<F>(&mut self, mut callback: F)
+    where
+        F: FnMut(&T, u32),
+    {
+        while let Some(curr_free_index) = self.next_free_slot() {
+            if let Some((value, occupied_index)) = self.next_occupied() {
+                callback(value, curr_free_index.0);
+                //The entry at curr_free_index.0 should have `None` by now.
+                //Moving it to `occupied_index` will make that entry empty.
+                self.elements.swap(curr_free_index.0, occupied_index);
+            } else {
+                //Could not find an occupied slot to fill the free slot
+                env::panic_str(ERR_INCONSISTENT_STATE)
+            }
+        }
+        self.elements.len = self.occupied_count;
```

Comment:
Inline with previous comment, modifying a type that is meant to be internal to `Vector` seems unideal since it seems like the same functionality could be done with the safe public APIs of `Vector`.

Reply:
I did not see a suitable API to meet the needs of `defrag`. If you have any specific suggestion I can look into it.